lenskit-slim/src/main/java/org/lenskit/slim/SLIMScoringStrategy.java

```diff
@@ -0,0 +1,143 @@
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.slim;
+
+import it.unimi.dsi.fastutil.longs.Long2DoubleMap;
+import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import org.grouplens.grapht.annotation.DefaultImplementation;
+import org.lenskit.util.math.Vectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.util.Map;
+
+import static org.lenskit.slim.LinearRegressionHelper.addVectors;
+
+
+/**
+ * Linear regression which minimize following loss function
+ * 1/2*||a_j - A*w_j||^2 + beta/2*||wj||^2 + lambda*||w_j||
+ *
+ * @author <a href="http://www.grouplens.org">GroupLens Research</a>
+ */
+@DefaultImplementation(CovarianceUpdate.class)
+public abstract class SLIMScoringStrategy {
+    protected static final Logger logger = LoggerFactory.getLogger(SLIMScoringStrategy.class);
+    protected final SLIMUpdateParameters updateParameters;
+
+    @Inject
+    public SLIMScoringStrategy(SLIMUpdateParameters updateParameters) {
+        this.updateParameters = updateParameters;
+    }
+
+    /**
+     * compute prediction of single observation
+     * @param observation test data
+     * @param weight weight
+     * @return prediction
+     */
+    public double computePrediction(Long2DoubleMap observation, Long2DoubleMap weight) {
+        return Vectors.dotProduct(weight, observation);
+    }
+
+    /**
+     * compute residual vector
+     *
+     * @param labels label vector
+     * @param dataMatrix column-wise data matrix
+     * @param weights weights vector
+     * @return residual vector
+     */
+    public Long2DoubleMap computeResiduals(Long2DoubleMap labels, Long2ObjectMap<Long2DoubleMap> dataMatrix, Long2DoubleMap weights) {
+        Long2DoubleMap predictions = predict(dataMatrix, weights);
+        return addVectors(labels, Vectors.multiplyScalar(predictions,-1.0));
+    }
+
+    /**
+     * learning process
+     *
+     * @param labels label vector
+     * @param trainingDataMatrix observations matrix row: user ratings for different items, column: item ratings of different users
+     * @return a trained weight vector
+     */
+    public abstract Long2DoubleMap fit(Long2DoubleMap labels, Long2ObjectMap<Long2DoubleMap> trainingDataMatrix);
```

Comment:
No. It is just kept as an alternative and used as a double check in the tests in case something goes wrong.